File: test/AuthZI.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraID.Common/Authorization/SimpleAuthorizationTests.fs
namespace AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common.Authorization

open AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common
open AuthZI.Tests.MicrosoftOrleans.Grains.SimpleAuthorization
open AuthZI.Extensions.TaskExtensions
open AuthZI.Security
open RootConfiguration
open System
open System.Security
open System.Threading.Tasks
open Xunit
open AuthZI.MicrosoftOrleans.Authorization

type SimpleAuthorizationTestsBase() =
  [<Theory>]
  [<MemberData(nameof (TestData.UserWithScopeAdeleV), MemberType = typeof<TestData>)>]
  member _.``An authenticated user can invoke the grain method``
    (userName: string)
    (password: string)
    (scope: string list)
    =
    task {
      // Arrange
      let! accessToken = AccessTokenProvider.getAccessTokenForUserOnWebClient1Async userName password

      let clusterClient = getClusterClient accessToken
      let simpleGrain = clusterClient.GetGrain<ISimpleGrain>(Guid.NewGuid())
      let! value = simpleGrain.GetWithAuthenticatedUser("Secret")

      Assert.True(value.Equals "Secret")
    }

  [<Theory>]
  [<MemberData(nameof (TestData.UserWithScopeAdeleV), MemberType = typeof<TestData>)>]
  member _.``An authenticated user on an unauthenticated client can't invoke the grain method``
    (userName: string)
    (password: string)
    (scope: string list)
    =
    task {
      // Arrange
      let! accessToken = AccessTokenProvider.getAccessTokenForUserOnWebClient2Async userName password

      let clusterClient = getClusterClient accessToken
      let simpleGrain = clusterClient.GetGrain<ISimpleGrain>(Guid.NewGuid())

      // Act
      let action =
        task {
          let! value = simpleGrain.GetValue()
          return value
        }

      let! result = Assert.ThrowsAsync<SecurityException>(fun () -> action)
      Assert.True(result.Message = SecureGrain.AccessDeniedMessage)
    }

  [<Fact>]
  member _.``An anonymous user can't invoke the grain method``() =
    task {
      // Arrange
      let accessToken = String.Empty

      let clusterClient = getClusterClient accessToken
      let simpleGrain = clusterClient.GetGrain<ISimpleGrain>(Guid.NewGuid())

      // Act
      let action =
        task {
          let! value = simpleGrain.GetWithAuthenticatedUser(String.Empty)
          return value
        }

      let! result = Assert.ThrowsAsync<InvalidOperationException>(fun () -> action)
      Assert.True(true)
    }

  [<Fact>]
  member _.``An anonymous user can invoke a grain method with [AllowAnonymous] attribyte``() =
    async {
      // Arrange
      let accessToken = String.Empty

      let clusterClient = getClusterClient accessToken
      let simpleGrain = clusterClient.GetGrain<ISimpleGrain>(Guid.Empty)

      let! value = simpleGrain.GetWithAnonymousUser("Secret") |> Async.AwaitTask

      Assert.True(value.Equals "Secret")
    }

File: test/AuthZI.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraID.Common.CSharp/Authorization/SimpleAuthorizationTests.cs
using AuthZI.MicrosoftOrleans.Authorization;
using AuthZI.Tests.MicrosoftOrleans.Grains.SimpleAuthorization;
using AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common.Initialization;
using System;
using System.Collections.Generic;
using System.Security;
using System.Threading.Tasks;
using Xunit;

namespace AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common.Authorization;

public class SimpleAuthorizationTestsBase(MainTestFixture fixture)
{
  [Theory]
  [MemberData(nameof(TestData.UserWithScopeAdeleV), MemberType = typeof(TestData), DisableDiscoveryEnumeration = true)]
  public async Task AnAuthenticatedUserCanInvokeTheGrainMethod(
    string userName,
    IEnumerable<string> scope)
  {
    var accessToken = await AccessTokenProvider.GetAccessTokenForUserOnWebClient1Async(userName);

    var clusterClient = fixture.GetClusterClient(accessToken);
    var simpleGrain = clusterClient.GetGrain<ISimpleGrain>(Guid.NewGuid());
    var value = await simpleGrain.GetWithAuthenticatedUser("Secret");

    Assert.True(value.Equals("Secret"));
  }

  [Theory]
  [MemberData(nameof(TestData.UserWithScopeAdeleV), MemberType = typeof(TestData), DisableDiscoveryEnumeration = true)]
  public async Task AnAuthenticatedUserOnAnUnauthenticatedClientCannotInvokeTheGrainMethod(
    string userName,
    IEnumerable<string> scope)
  {
    var accessToken = await AccessTokenProvider.GetAccessTokenForUserOnWebClient2Async(userName);

    var clusterClient = fixture.GetClusterClient(accessToken);
    var simpleGrain = clusterClient.GetGrain<ISimpleGrain>(Guid.NewGuid());

    var result = await Assert.ThrowsAsync<SecurityException>(() => simpleGrain.GetValue());
    Assert.True(result.Message == SecureGrain.AccessDeniedMessage);
  }

  [Fact]
  public async Task AnAnonymousUserCannotInvokeTheGrainMethod()
  {
    var accessToken = string.Empty;

    var clusterClient = fixture.GetClusterClient(accessToken);
    var simpleGrain = clusterClient.GetGrain<ISimpleGrain>(Guid.NewGuid());

    await Assert.ThrowsAsync<InvalidOperationException>(() => simpleGrain.GetWithAuthenticatedUser(string.Empty));
    Assert.True(true);
  }

  [Fact]
  public async Task AnAnonymousUserCanInvokeAGrainMethodWithAllowAnonymousAttribute()
  {
    var accessToken = string.Empty;

    var clusterClient = fixture.GetClusterClient(accessToken);
    var simpleGrain = clusterClient.GetGrain<ISimpleGrain>(Guid.Empty);
    var value = await simpleGrain.GetWithAnonymousUser("Secret");

    Assert.True(value.Equals("Secret"));
  }
}
